### PR TITLE
Adding surrealDB module

### PR DESCRIPTION
**Pull request by jespino**
## What does this PR do?

This PR adds the SurrealDB module. (This module is heavily inspired by the rust implementation)

## Why is it important?

SurrealDB is a database system that is getting more a more popular lately, so it can help users to test it using testcontainers.

## How to test this PR

This PR include unit tests for the module

### CODE REVIEW

File: modules/surrealdb/examples_test.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+	surrealdbContainer, err := surrealdb.RunContainer(ctx, testcontainers.WithImage("surrealdb/surrealdb:v1.1.1"))
+	if err != nil {
+		panic(err)
+	}
+
+	// Clean up the container
+	defer func() {
+		if err := surrealdbContainer.Terminate(ctx); err != nil {
+			panic(err)
```
Comment (mdelapenya) [inline modules/surrealdb/examples_test.go:23]: In #2193 we are replacing panics with log.Fatalf. Could you proceed as in that PR with this module? I'll probably merge that PR first, as it replaces all the panics in testable examples with log.Fatal

Comment (jespino) [reply modules/surrealdb/examples_test.go]: Done!

**Comment by mdelapenya** [comment]
Same as https://github.com/testcontainers/testcontainers-go/pull/2244#issuecomment-1951928524. Would you evaluate to host the module? As described in the docs, if you don't feel can do it, it's fine having it here

**Comment by jespino** [comment]
@mdelapenya for this specific one I prefer it to go here.

**Comment by mdelapenya** [comment]
I think the lint is complaining on the module https://github.com/testcontainers/testcontainers-go/actions/runs/8018977780/job/21906230055?pr=2192

Could you run `make lint` and verify the errors?

**Comment by jespino** [comment]
@mdelapenya done

File: modules/surrealdb/surrealdb.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+// It is used in conjunction with WithUsername to set a username and its password.
+// It will set the superuser password for MongoDB.
+func WithPassword(password string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Env["SURREAL_PASS"] = password
+	}
+}
+
+// WithAuthentication enables or disables authentication for the SurrealDB instance
+func WithAuthentication(enabled bool) testcontainers.CustomizeRequestOption {
```
Comment (mdelapenya) [inline modules/surrealdb/surrealdb.go:52]: If the default is auth=true, then I'd create an option only for disabling it, something like `WithoutAuthentication` or `NoAuthentication`.

Thoughts?

File: modules/surrealdb/surrealdb.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		if enabled {
+			req.Env["SURREAL_AUTH"] = "true"
+		} else {
+			req.Env["SURREAL_AUTH"] = "false"
+		}
+	}
+}
+
+// WithStrict enables or disables strict mode for the SurrealDB instance
+func WithStrict(enabled bool) testcontainers.CustomizeRequestOption {
```
Comment (mdelapenya) [inline modules/surrealdb/surrealdb.go:63]: Same here, if the module is going to set a default, I'd let the option to set the opposite value.

Also, could you include a test calling it?

File: modules/surrealdb/surrealdb.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		if enabled {
+			req.Env["SURREAL_STRICT"] = "true"
+		} else {
+			req.Env["SURREAL_STRICT"] = "false"
+		}
+	}
+}
+
+// WithAllowAllCaps enables or disables all caps for the SurrealDB instance
+func WithAllowAllCaps(enabled bool) testcontainers.CustomizeRequestOption {
```
Comment (mdelapenya) [inline modules/surrealdb/surrealdb.go:74]: Same here

**Review by mdelapenya — COMMENTED** [review]
@jespino looks good to me having the module here, let's address the comments and we are to go. Thank you for your time 🙇

File: modules/surrealdb/surrealdb.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// WithAuthentication enables authentication for the SurrealDB instance
+func WithAuthentication() testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Env["SURREAL_AUTH"] = "true"
+	}
+}
+
+// WithStrict enables or disables strict mode for the SurrealDB instance
```
Comment (mdelapenya) [inline modules/surrealdb/surrealdb.go:58]:
```suggestion
// WithStrict enables strict mode for the SurrealDB instance
```

File: modules/surrealdb/surrealdb.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+func WithStrictMode() testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Env["SURREAL_STRICT"] = "true"
+	}
+}
+
+// WithAllowAllCaps enables or disables all caps for the SurrealDB instance
```
Comment (mdelapenya) [inline modules/surrealdb/surrealdb.go:65]:
```suggestion
// WithAllowAllCaps enables all caps for the SurrealDB instance
```

File: docs/modules/surrealdb.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+When starting the SurrealDB container, you can pass options in a variadic way to configure it.
+
+#### Image
+
+If you need to set a different SurrealDB Docker image, you can use `testcontainers.WithImage` with a valid Docker image
+for SurrealDB. E.g. `testcontainers.WithImage("surrealdb/surrealdb:v1.1.1")`.
+
+{% include "../features/common_functional_options.md" %}
+
```
Comment (mdelapenya) [inline docs/modules/surrealdb.md:44]: Let's include here docs for the functional opts for the module: `WithStrict` and friends...

**Review by mdelapenya — COMMENTED** [review]
It LGTM although I added a few comments 🙏 

Thanks!

**Comment by mdelapenya** [comment]
@jespino I added the docs for the module, thanks! Please take a look before merging if possible 🙏

File: docs/modules/surrealdb.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+#### Set username and password
+
+If you need to set a different database, and its credentials, you can use `WithUsername` and `WithPassword` options.
```
Comment (jespino) [inline docs/modules/surrealdb.md:47]: I don't understand why here it says, "set a different database", it is setting the username and the password, but not the database.

Comment (mdelapenya) [reply docs/modules/surrealdb.md:47]: Oh oh sorry, wrong copy paste. My bad. Updating it now

Comment (mdelapenya) [inline docs/modules/surrealdb.md:47]:
```suggestion
If you need to set different credentials, you can use `WithUsername` and `WithPassword` options.
```

File: docs/modules/surrealdb.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+!!!info
+    The default values for the username is `root`, for password is `test` and for the default database name is `test`.
```
Comment (jespino) [inline docs/modules/surrealdb.md:50]: The default password is `root`, not `test`. And there is no "default database". The databases are created on the fly, AFAIK from playing around with it a bit.

**Comment by jespino** [comment]
@mdelapenya sorry about the delay, I've been pretty busy lately. I added a couple of comments (one of them you already fixed it). Otherwise, it looks good to me. Thanks!

File: docs/modules/surrealdb.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+for SurrealDB. E.g. `testcontainers.WithImage("surrealdb/surrealdb:v1.1.1")`.
+
+{% include "../features/common_functional_options.md" %}
+
+#### Set username and password
+
+If you need to set different credentials, you can use `WithUsername` and `WithPassword` options.
+
+!!!info
+    The default values for the username is `root`, for password is `test` and for the default database name is `test`.
```
Comment (mdelapenya) [inline docs/modules/surrealdb.md:50]:
```suggestion
    The default values for the username and the password is `root`.
```

**Comment by mdelapenya** [comment]
@jespino I'm merging this one. Thank you so much for your contributions, they are adding lots of value to the project!!

**Review by mdelapenya — DISMISSED** [review]
LGTM

**Pull request merged**